### PR TITLE
Bug fix/supervision server cleanup

### PR DESCRIPTION
**Pull request by kvahed**
# Pull Request Guidelines

Pull requests are an essential collaborative tool for modern software development. 

The below list is intended to help you figure out whether your code is ready to be reviewed
and merged into ArangoDB. The overarching goal is to:

- Reduce the amount of *recurring* defects
- Reduce the impact to the other developer’s time and energy spent on defects in your code
- Increase the overall autonomy and productivity of individual developers

## Acceptance Checklist

*The below list is not exhaustive, think thoroughly whether the provided information is sufficient.*
*Remove options that do not apply*

### Scope & Purpose

*(Can you describe what functional change your PR is trying to effect?)*

- [ ] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [ ] Bug-Fix for a *released version* (did you remember to port this to all relevant release branches?)
- [ ] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [ ] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [ ] The behavior change can be verified via automatic tests

#### Related Information

*(Please reference tickets / specification etc )*

- [ ] There is a GitHub Issue reported by a Community User: 
- [ ] There is an internal planning ticket: 
- [ ] There is a *JIRA Ticket number* (In case a customer was affected / involved): 
- [ ] There is a design document: 

### Testing & Verification

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This PR adds tests that were used to verify all changes:

- [ ] Added **Regression Tests** (Only for bug-fixes) 
- [ ] Added new C++ **Unit Tests** (Either GoogleTest or Catch-Test)
   - Did you add tests for a new *RestHandler* subclass ?
   - Did you add new mocks of underlying code layers to be able to verify your functionality ? 
   - ...
- [ ] Added new **integration tests** (i.e. in shell_server / shell_server_aql)
- [ ] Added new **resilience tests** (only if the feature is impacted by failovers)

Additionally:

- [ ] There are tests in an external testing repository (i.e. node-resilience tests, chaos tests)
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

*(Include link to Jenkins run etc)*

> Think about whether the new code you added is modular enough to be
> easily testable by unit tests written with GTest / Catch. It is not good if your feature is so interconnected
> that it prevents other people from writing their own unit gests. It should be possible
> to use your code in future without extensively mocking your classes.
> A bad example that required some extensive effort would be the storage engine API.

### Documentation

> All new Features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the changelog so that
> developers and users have a concise overview. 

- [ ] Added a *Changelog Entry* (referencing the corresponding public or internal issue number)
- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *http API* 
- [ ] Added *Swagger examples* for the http API  

### CLA Note 

Please note that for legal reasons we require you to sign the [Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

File: CHANGELOG [hunk 1]
```diff
@@ -16,15 +19,15 @@ devel
...
   network roundtrips have a higher share of the total query execution times there.
 
 * Traversal performance can also be improved by not fetching the visited vertices
-  from the storage engine in case the traversal query does not refer to them. 
+  from the storage engine in case the traversal query does not refer to them.
   For example, in the query
 
       FOR v, e, p IN 1..3 OUTBOUND 'collection/startVertex' edges
-        RETURN e
+	RETURN e
```
Comment (ObiWahn) [inline CHANGELOG]: This breaks the formatting of the CHANGELOG.

File: arangod/Agency/Supervision.cpp [hunk 2]
```diff
@@ -918,8 +948,8 @@ void Supervision::run() {
 // Guarded by caller
 bool Supervision::isShuttingDown() {
   _lock.assertLockedByCurrentThread();
-  if (_snapshot.has("Shutdown")) {
-    return _snapshot.hasAsBool("Shutdown").first;
+  if ((*_snapshot).has("Shutdown")) {
```
Comment (ObiWahn) [inline arangod/Agency/Supervision.cpp]:
```suggestion
  if ( _snapshot && _snapshot->has("Shutdown")) {
```

File: arangod/Agency/Supervision.h [hunk 3]
```diff
@@ -223,7 +226,8 @@ class Supervision : public arangodb::CriticalThread {
 
   Mutex _lock;   // guards snapshot, _jobId, jobIdMax, _selfShutdown
   Agent* _agent; /**< @brief My agent */
-  Node _snapshot;
+  Store _spearhead;
+  std::shared_ptr<Node const> _snapshot;
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Agency/Supervision.h]: fixed

File: arangod/Agency/Supervision.cpp [hunk 4]
```diff
@@ -851,8 +867,8 @@ void Supervision::run() {
     _agent->executeLockedRead([&]() {
       if (_agent->readDB().has(supervisionNode)) {
         try {
-          _snapshot = _agent->readDB().get(supervisionNode);
-          if (_snapshot.children().size() > 0) {
+          auto const sn = _agent->readDB().get(supervisionNode);
```
Comment (neunhoef) [inline arangod/Agency/Supervision.cpp]: This is a change in behaviour. _snapshot is no longer initialized here. Is this safe?

File: arangod/Agency/Supervision.cpp [hunk 5]
```diff
@@ -882,7 +898,7 @@ void Supervision::run() {
       {
         MUTEX_LOCKER(locker, _lock);
 
-        if (isShuttingDown()) {
+        if (_snapshot != nullptr && isShuttingDown()) {
```
Comment (neunhoef) [inline arangod/Agency/Supervision.cpp]: This is a change in behaviour. Why is this safe?

File: arangod/Agency/Supervision.cpp [hunk 6]
```diff
@@ -1009,8 +1038,8 @@ void Supervision::run() {
 // Guarded by caller
 bool Supervision::isShuttingDown() {
   _lock.assertLockedByCurrentThread();
-  if (_snapshot.has("Shutdown")) {
-    return _snapshot.hasAsBool("Shutdown").first;
+  if ( _snapshot && _snapshot->has("Shutdown")) {
```
Comment (neunhoef) [inline arangod/Agency/Supervision.cpp]: Why do we check in some places if `_snapshot` is non-nullptr bui not in all places?

File: arangod/Agency/Supervision.cpp [hunk 7]
```diff
@@ -1079,6 +1108,163 @@ void Supervision::handleShutdown() {
...
+            persistedTimeStamp.clear();
+          }
+        }
+      } catch (std::exception const& e) {
+        LOG_TOPIC("21a9e", DEBUG, Logger::SUPERVISION)
+          << "Failing to analyse " << serverId << " as deletion candidate " << e.what();
+      }
+
+      // We are still here?
+      serverList.emplace(std::pair<ServerID, std::string>{serverId.first, persistedTimeStamp});
```
Comment (neunhoef) [inline arangod/Agency/Supervision.cpp]: This is suboptimal: A pair is constructed and then copy-constructed in the container. I suggest to leave out the explicit `std::pair` construction.

File: arangod/Agency/Supervision.cpp [hunk 8]
```diff
@@ -1079,6 +1108,163 @@ void Supervision::handleShutdown() {
...
+        }
+        if (!server.second.empty()) { // Timestamp unchanged only, if persistent entry
+          VPackObjectBuilder prec(&trxs);
+          trxs.add("/arango/Supervision/Health/" + serverName + "/Timestamp", VPackValue(server.second));
+        }
+      }
+    }
+  }
+  if (servers.size() > 0) {
+    _nextServerCleanup = std::chrono::system_clock::now() + std::chrono::seconds(3600);
```
Comment (neunhoef) [inline arangod/Agency/Supervision.cpp]: Should this not happen outside of the if? If nothing is deleted this looks like a busy look eventually.

**Comment by neunhoef** [comment]
And: we should briefly talk about this tomorrow: How is it guaranteed, that `_snapshot` in `Supervision` is always non-nullptr if it is dereferenced?

File: arangod/Agency/Supervision.cpp (hunk 7 above)
Comment (ObiWahn) [reply arangod/Agency/Supervision.cpp]: How about `try_emplace` or `std::piecewise_construct`

**Pull request merged**